src/components/fields/Percentage/Settings.tsx

```diff
@@ -0,0 +1,186 @@
+import { useEffect, useState } from "react";
+
+import {
+  Box,
+  Checkbox,
+  Grid,
+  InputLabel,
+  MenuItem,
+  TextField,
+  Typography,
+  useTheme,
+} from "@mui/material";
+import ColorPickerInput from "@src/components/ColorPickerInput";
+import { ISettingsProps } from "@src/components/fields/types";
+
+import { Color, toColor } from "react-color-palette";
+import { fieldSx } from "@src/components/SideDrawer/utils";
+import { resultColorsScale, defaultColors } from "@src/utils/color";
+import { GlobalStyles } from "tss-react";
+
+const colorLabels: { [key: string]: string } = {
+  0: "Start",
+  1: "Middle",
+  2: "End",
+};
+
+export default function Settings({ onChange, config }: ISettingsProps) {
+  const [colorsDraft, setColorsDraft] = useState<any[]>(
+    config.colors ? config.colors : defaultColors
+  );
+
+  const [checkStates, setCheckStates] = useState<boolean[]>([
+    colorsDraft[0],
+    colorsDraft[1],
+    colorsDraft[2],
+  ]);
+
+  useEffect(() => {
+    onChange("colors")(colorsDraft);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [colorsDraft]);
+
+  const onCheckboxChange = (index: number, checked: boolean) => {
+    setColorsDraft(
+      colorsDraft.map((value: any, idx: number) =>
+        index === idx ? (checked ? value || defaultColors[idx] : null) : value
+      )
+    );
+    setCheckStates(
+      checkStates.map((value, idx) => (index === idx ? checked : value))
+    );
+  };
+
+  const handleColorChange = (index: number, color: Color): void => {
+    setColorsDraft(
+      colorsDraft.map((value, idx) => (index === idx ? color.hex : value))
+    );
+  };
```

Comment:
We’re getting into the weeds of React here—I don’t think you need to make a `colorsDraft` state, since that state definition only ensures that it’s not undefined and provides a default value.

You then sync up `colorsDraft` with the parent component using an Effect that calls `onChange`. This also isn’t necessary—you can read more about that in the [React docs](https://beta.reactjs.org/learn/you-might-not-need-an-effect#notifying-parent-components-about-state-changes). Disabling this eslint rule is also a code smell: `// eslint-disable-next-line react-hooks/exhaustive-deps`.

Instead, you could ensure the `colorsDraft` is a correct value without state:
```ts
const colors = config.colors ?? defaultColors;
```

Then, call `onChange` directly in your change handlers, so you’re only ever updating the state in the parent component [`ColumnConfig`](https://github.com/rowyio/rowy/blob/0b81f80e6cc16c324f92b2b2ceec3e7c72d227b2/src/components/ColumnModals/ColumnConfigModal/ColumnConfig.tsx#L43).

```suggestion
  const colors = config.colors ?? defaultColors;

  // It makes sense to keep this a state
  const [checkStates, setCheckStates] = useState<boolean[]>([
    colorsDraft[0],
    colorsDraft[1],
    colorsDraft[2],
  ]);

  const onCheckboxChange = (index: number, checked: boolean) => {
    onChange("colors")(
      colorsDraft.map((value: any, idx: number) =>
        index === idx ? (checked ? value || defaultColors[idx] : null) : value
      )
    );
    setCheckStates(
      checkStates.map((value, idx) => (index === idx ? checked : value))
    );
  };

  const handleColorChange = (index: number, color: Color): void => {
    onChange("colors")(
      colorsDraft.map((value, idx) => (index === idx ? color.hex : value))
    );
  };
```

Reply:
Derived states just bring simplicity... 💯

Reply:
I've said to you that I read this article twice. What a shame 👀